src/jpegxl.imageio/jxloutput.cpp

```diff
@@ -165,23 +192,72 @@ JxlOutput::open(const std::string& name, const ImageSpec& newspec,
 
     m_frame_settings = JxlEncoderFrameSettingsCreate(m_encoder.get(), nullptr);
 
-    // const float quality = 100.0;
-    const int effort = 7;
-    const int tier   = 0;
-    // Lossless only makes sense for integer modes
-    if (m_basic_info.exponent_bits_per_sample == 0) {
-        // Must preserve original profile for lossless mode
-        m_basic_info.uses_original_profile = JXL_TRUE;
-        JxlEncoderSetFrameDistance(m_frame_settings, 0.0);
-        JxlEncoderSetFrameLossless(m_frame_settings, JXL_TRUE);
+    // JpegXL Compression Settings
+
+    bool lossless = true;
+
+    // Distance Mutually exclusive with quality
+    if (m_spec.find_attribute("jpegxl:distance")) {
+        const float distance = m_spec.get_float_attribute("jpegxl:distance",
+                                                          0.0f);
+
+        m_basic_info.uses_original_profile = distance == 0.0f ? JXL_TRUE
+                                                              : JXL_FALSE;
+        JxlEncoderSetFrameDistance(m_frame_settings, distance);
+        JxlEncoderSetFrameLossless(m_frame_settings,
+                                   distance == 0.0f ? JXL_TRUE : JXL_FALSE);
+
+        lossless = distance == 0.0f;
+
+        DBG std::cout << "Compression distance set to " << distance << "\n";
+
+    } else {
+        auto compqual = m_spec.decode_compression_metadata("jpegxl", 100);
+        if (Strutil::iequals(compqual.first, "jpegxl")) {
+            if (compqual.second == 100) {
+                m_basic_info.uses_original_profile = JXL_TRUE;
+                JxlEncoderSetFrameDistance(m_frame_settings, 0.0);
+                JxlEncoderSetFrameLossless(m_frame_settings, JXL_TRUE);
+
+                lossless = true;
+            } else {
+                m_basic_info.uses_original_profile = JXL_FALSE;
+                JxlEncoderSetFrameDistance(
+                    m_frame_settings,
+                    1.0f / static_cast<float>(compqual.second));
+                JxlEncoderSetFrameLossless(m_frame_settings, JXL_FALSE);
+            }
+        } else {  // default to lossless
```

Comment:
Is this the right default? (I don't know, really am asking.) For no compression request, you use "jpegxl:100", and revert to lossless for any unrecognized compression name.

Imagine `oiiotool foo.xyz -o foo.jxl` so you may write with some compression name from another format without the user really being aware that this is making them lapse into lossless compression for the jxl file.

A possible alternative would be to use lossless if the compression was "none" or "lossless", but any other *unrecognized* name would default to a reasonable level of jpegxl compression (as you did if no "compression" request was made at all).

You be the judge -- I don't know what JXL user expectations would be, nor do I know how big a compression or quality different there will be between the lossless mode and the jpegxl:100 default you use when no compression is specified.

Reply:
Personally i love Jpeg2000 and JpegXL because they have lossless compression that in same time doing files smaller than any other lossless file formats.
But if you think for a users better to set same settings as used in cjxl.exe from libjxl that use probably 70 or 95 (don't remember exactly)

Reply:
No, I trust your judgment. The information I didn't have is whether the lossless compression was any good. I'm ok with the rule of "you get lossless unless you specifically ask for lossy by name" as long as the lossless compression is pretty good and will not seem like a strange choice to JXL users.

Reply:
The reference implementation's cjxl losslessly transcodes jpeg inputs and uses normal lossless on GIF inputs, with a lossy quality of 90 by default for anything else

Reply:
@jonnyawsom3, Maybe that makes sense for a cxl tool that is an image re-encoding tool, and encodings is a "goal"?
When OpenImageIO is a library that is an image IO interface and data are the "hero"?

@lgritz uint16:
- tiff (zip) 7.93Mb
- png (9) 5.48Mb
- jxl (lossless) 2.86Mb

half:
- exr (piz) 3.71mb
- jxl (lossless) 2.28mb

And in same time jxl support hdr, multichannels, layers, etc. and sound like it can be read on any refrigerator soon :)